core/src/cell.rs

```diff
@@ -2,39 +2,152 @@ use crate::block::Block;
 use crate::header::Header;
 use crate::transaction::{CellOutPoint, CellOutput, OutPoint, Transaction};
 use crate::Capacity;
+use crate::{BlockNumber, EpochNumber};
 use ckb_util::LowerHexOption;
 use fnv::{FnvHashMap, FnvHashSet};
 use numext_fixed_hash::H256;
 use serde_derive::{Deserialize, Serialize};
 use std::fmt;
 
+#[derive(Clone, Debug, Eq, PartialEq, Default, Deserialize, Serialize)]
+pub struct BlockInfo {
+    pub number: BlockNumber,
+    pub epoch: EpochNumber,
+}
+
 #[derive(Clone, Eq, PartialEq, Default, Deserialize, Serialize)]
 pub struct CellMeta {
     #[serde(skip)]
     pub cell_output: Option<CellOutput>,
     pub out_point: CellOutPoint,
-    pub block_number: Option<u64>,
+    pub block_info: Option<BlockInfo>,
     pub cellbase: bool,
     pub capacity: Capacity,
     pub data_hash: Option<H256>,
 }
 
-impl From<&CellOutput> for CellMeta {
-    fn from(output: &CellOutput) -> Self {
+#[derive(Default)]
+pub struct CellMetaBuilder {
+    cell_output: Option<CellOutput>,
+    out_point: CellOutPoint,
+    block_info: Option<BlockInfo>,
+    cellbase: bool,
+    capacity: Capacity,
+    data_hash: Option<H256>,
+}
+
+impl CellMetaBuilder {
+    pub fn from_cell_meta(cell_meta: CellMeta) -> Self {
+        let CellMeta {
+            cell_output,
+            out_point,
+            block_info,
+            cellbase,
+            capacity,
+            data_hash,
+        } = cell_meta;
+        Self {
+            cell_output,
+            out_point,
+            block_info,
+            cellbase,
+            capacity,
+            data_hash,
+        }
+    }
+
+    pub fn from_cell_output(cell_output: CellOutput) -> Self {
+        CellMetaBuilder::default()
+            .capacity(cell_output.capacity)
+            .cell_output(cell_output)
+    }
+
+    pub fn cell_output(mut self, cell_output: CellOutput) -> Self {
+        self.cell_output = Some(cell_output);
+        self
+    }
+
+    pub fn out_point(mut self, out_point: CellOutPoint) -> Self {
+        self.out_point = out_point;
+        self
+    }
+
+    pub fn block_info(mut self, block_info: BlockInfo) -> Self {
+        self.block_info = Some(block_info);
+        self
+    }
+
+    pub fn block_number(mut self, block_number: BlockNumber) -> Self {
+        if let Some(ref mut block_info) = self.block_info {
+            block_info.number = block_number;
+        } else {
+            self.block_info = Some(BlockInfo {
+                number: block_number,
+                epoch: 0,
+            });
+        }
+        self
+    }
+
+    pub fn epoch_number(mut self, epoch_number: EpochNumber) -> Self {
+        if let Some(ref mut block_info) = self.block_info {
+            block_info.epoch = epoch_number;
+        } else {
+            self.block_info = Some(BlockInfo {
+                number: 0,
+                epoch: epoch_number,
+            });
+        }
+        self
+    }
```

Comment:
Can we remove `epoch_number` `block_number` from the `CellMetaBuilder`?  
Just keep `block_info`,  add some constructor on BlockInfo for test convenient.